docs/proposals/dynamic-kubelet-configuration.md

```diff
@@ -0,0 +1,290 @@
+  + The Kubelet will verify the downloaded `ConfigMap` by performing this same procedure and comparing the result to the hash in the name. This helps prevent the "shoot yourself in the foot" scenario detailed below in *Operational Considerations/Rollout workflow*.
+
+`ConfigMap` object containing just the Kubelet's configuration:
+```
+kind: ConfigMap
+metadata:
+  name: node-config-{sha1 hash of ConfigMap, sans identifying information, for verification}
+data:
+  kubelet: "{JSON blob}"
+```
+
+#### On-disk
+
+- The Kubelet should accept a `--config-dir` flag that specifies a directory (e.g. `config-dir`).
+- The Kubelet will use this directory to checkpoint downloaded `ConfigMap`s.
+  + When the Kubelet downloads a `ConfigMap`, it will save its contents to a directory named after the UID of that `ConfigMap`, in a subdirectory of `config-dir` called `configmaps`, e.g. `config-dir/configmaps/{uid}`.
+  + The directory `config-dir/configmaps/{uid}` contains the `Data` of the corresponding `ConfigMap` as a set of files. Given the most recent `ConfigMap` example above, reading `config-dir/configmaps/{uid-of-the-example}/kubelet` would return `{JSON blob}`.
+- The user can additionally pre-populate `config-dir/init` with an initial set of configuration files, to be used prior to the Kubelet being able to access the API server (including when running in standalone mode).
+
+### Orchestration of configuration
+
+There are a lot of opinions around how to orchestrate configuration in a cluster. The following items form the basis of a robust solution:
+
+#### Robust Kubelet behavior
+
+To make config updates robust, the Kubelet should be able to locally and automatically recover from bad config pushes. We should strive to avoid requiring operator intervention, though this may not be possible in all scenarios.
+
+Recovery involves:
+- Checkpointing configuration on-disk, so prior versions are locally available for rollback.
+- Tracking a LKG (last-known-good) configuration, which will be the rollback target if the current configuration turns out to be bad.
+- Tracking the health of the Kubelet against a given configuration and remembering, at least for a time, if a certain configuration correlated with poor health.
+- Rolling back to LKG if the current configuration resulted in poor health.
+- Providing operators an escape hatch for exiting dead-end states.
+
+##### Finding and checkpointing intended configuration
+
+The Kubelet finds its intended configuration by looking for the `ConfigMap` referenced via it's `Node`'s `spec.config` field. This field is an `ObjectReference`. We consider this field to be "empty" if it lacks the information necessary to resolve to an object. The field must provide at least `name` or `uid`, to be considered "non-empty." If no `namespace` or `uid` is provided, the `default` namespace is assumed. 
```

Comment:
I modified it to use the wrapper type, require namespace, and separate empty from invalid, let me know what you think.